### ISSUE TITLE
--data-dir is ignored in defaults files' search path 

### ISSUE THREAD

**Issue by tstenner** [jgm/pandoc#7320]
With pandoc 2.13, the data dir is used to the search path for default files supplied with the `--defaults` parameter but not in other default files:

``` bash
$ mkdir -p test/defaults
$ printf 'to: markdown' > test/defaults/a.yaml
$ printf 'defaults: a' > test/defaults/b.yaml
$ echo 'Test' | pandoc --data-dir=test --defaults=a -  # loads test/defaults/a.yaml
Test
$ echo 'Test' | pandoc --data-dir=test --defaults=b -  # loads test/defaults/b.yaml, but doesn't find a.yaml in the same directory
pandoc: a.yaml: openBinaryFile: does not exist (No such file or directory)
$ cd test/defaults
$ echo 'Test' | pandoc --defaults=b -   # a.yaml in the working directory is found
Test
```

**Comment by jgm** [comment]
See https://github.com/jgm/pandoc/pull/6924
which explains the motivation for the way it is now (under "user data directories").
I'm not sure about this.

**Comment by tstenner** [comment]
`defsA` not inheriting the data dir from `defsB` is a bit surprising, but I can see that there are valid reasons for it.

Both files are in the same directory, so I'd expect that if `b.yaml` is found, so should `a.yaml`.
Also, when I put both files in the default `--data-dir` path (`~/.local/share/pandoc/defaults`), it works just fine.

So, for `defaults: a`, pandoc searches for `a.yaml` in the `defaults` subdirectory of `data-dir:` (defaults to the user data dir) or the current working directory. I can make it work by putting

``` yaml
data-dir: .
defaults: a
```

in `b.yaml`, but it's still unexpected.

**Comment by tstenner** [comment]
Some more rough edges: both `data-dir:` and `defaults:` don't interpolate the environment variables `${.}` etc., so neither

```
data-dir: ${.}
defaults: a
```

nor 

```
defaults: ${.}/a.yaml
```

includes a.yaml.

**Comment by jgm** [comment]
I wonder if @davidmrt98 (who implemented the code allowing default files to be specified within default files) has a comment on this issue?

**Comment by jgm** [comment]
On the second "rough edge," judging from the code, `data-dir` should accept the interpolation syntax.
I just tried it and confirmed that it works.

`defaults` does not currently accept the interpolation syntax.  It should, probably.

**Comment by tstenner** [comment]
> On the second "rough edge," judging from the code, data-dir should accept the interpolation syntax.
I just tried it and confirmed that it works.

I just checked and it doesn't:

``` bash
$ pandoc --version
pandoc 2.13 [..]
$ mkdir -p testdir/defaults
$ printf 'to: markdown' > testdir/defaults/a.yaml
$ printf 'data-dir: ${USERDATA}\ndefaults: a' > testdir/defaults/b.yaml
$ echo '# test' | pandoc --data-dir=testdir -d b -
pandoc: a.yaml: openBinaryFile: does not exist (No such file or directory)
$ printf 'data-dir: ${.}\ndefaults: a' > testdir/defaults/b.yaml
$ echo '# test' | pandoc --data-dir=testdir -d b -
pandoc: a.yaml: openBinaryFile: does not exist (No such file or directory)
$ printf 'data-dir: .\ndefaults: a' > testdir/defaults/dot.yaml
echo '# test' | pandoc --data-dir=testdir -d dot -
```

But for some reason, this does:

```
$ cd testdir/
$ echo '# test' | pandoc --data-dir=. -d dot -
# test
```
```

**Comment by jgm** [comment]
I'm not happy with the complexity of the code that handles this; I may try to rewrite it in a simpler way.

**Comment by davidmrt98** [comment]
I agree that the behaviour in the first example is odd. I already have a suspicion about the cause, but I will take a closer look.

On the interpolation issue I can not comment, as I was not involved in the implementation of this feature.

**Comment by 7596ff** [comment]
I'm currently building something that is affected by this issue, and I wanted to see if I had the current status of it correct.

There are in fact two bugs related to code in #6924: 
- Per the [docs](https://pandoc.org/MANUAL.html#option--defaults), A defaults file specified with `--defaults` is searched for in the current working directory and the `defaults` subdirectory of the data directory. If that defaults file includes other defaults files with `defaults:` per defaults file inheritance implemented, a search for them will not inherit the `data-dir` value passed from either the command line or the first defaults file itself.
- In defaults files, `${.}` is not interpreted correctly.

Similar problems have been brought up in the following issues: #8024 #8780 #8826 #8841

As for the first problem, in my project, I can work around it by programmatically inserting the correct data-dir in each defaults file, but this is not an ideal solution.

I don't want to ask for any guarantees or ETAs, but is this an accurate summation of the status of this problem?

**Comment by mboyea** [comment]
> `defaults` does not currently accept the interpolation syntax. It should, probably.

This would be an extremely useful feature for me.
Judging by the half-dozen issues about the feature, I think it would be useful for others too.
When I find time, I may contribute towards implementing this feature.

![image](https://github.com/jgm/pandoc/assets/68197922/2253c4a9-a5cb-4844-bea8-1df6b2f4fe25)

**Comment by jarnowic** [comment]
> Some more rough edges: both `data-dir:` and `defaults:` don't interpolate the environment variables `${.}` etc., so neither
> 
> ```
> data-dir: ${.}
> defaults: a
> ```
> 
> nor
> 
> ```
> defaults: ${.}/a.yaml
> ```
> 
> includes a.yaml.

I fear those environment variables won't be acceptable on Windows, so it seems tome that it is good that they are not considered: I am primarily a Windows user...
